File: src/main/java/com/rokucraft/rokubot/command/commands/RootTagCommand.java
package com.rokucraft.rokubot.command.commands;

import com.rokucraft.rokubot.command.GuildIndependentCommand;
import com.rokucraft.rokubot.command.SlashCommand;
import com.rokucraft.rokubot.entities.Tag;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.build.CommandData;
import net.dv8tion.jda.api.interactions.commands.build.Commands;
import net.dv8tion.jda.api.utils.messages.MessageCreateData;
import org.checkerframework.checker.nullness.qual.NonNull;

import static net.dv8tion.jda.api.interactions.commands.DefaultMemberPermissions.DISABLED;
import static net.dv8tion.jda.api.interactions.commands.DefaultMemberPermissions.ENABLED;

public class RootTagCommand implements GuildIndependentCommand, SlashCommand {
    private final @NonNull CommandData data;
    private final @NonNull MessageCreateData message;

    public RootTagCommand(@NonNull Tag tag, boolean defaultEnabled) {
        this.data = Commands.slash(tag.name(), (tag.description() != null) ? tag.description() : tag.name())
                .setDefaultPermissions(defaultEnabled ? ENABLED : DISABLED);
        this.message = tag.message();
    }

    public RootTagCommand(@NonNull Tag tag) {
        this(tag, true);
    }

    @Override
    public void execute(@NonNull SlashCommandInteractionEvent event) {
        event.reply(message).queue();
    }

    @Override
    public @NonNull CommandData getData() {
        return data;
    }
}

File: src/main/java/com/rokucraft/rokubot/command/commands/RootTagCommand.kt
package com.rokucraft.rokubot.command.commands

import com.rokucraft.rokubot.command.GuildIndependentCommand
import com.rokucraft.rokubot.command.SlashCommand
import com.rokucraft.rokubot.entities.Tag
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.commands.DefaultMemberPermissions
import net.dv8tion.jda.api.interactions.commands.build.CommandData
import net.dv8tion.jda.api.interactions.commands.build.Commands

class RootTagCommand @JvmOverloads constructor(
    private val tag: Tag,
    private val defaultPermissions: DefaultMemberPermissions = DefaultMemberPermissions.DISABLED
) : GuildIndependentCommand, SlashCommand {
    override fun execute(event: SlashCommandInteractionEvent) {
        event.reply(tag.message).queue()
    }

    override fun getData(): CommandData {
        return Commands.slash(tag.name, tag.description ?: tag.name)
            .setDefaultPermissions(defaultPermissions)
    }
}
